File: ExtraCredit/HW7/HW7.py
from typing import List

def minMeetingRooms(self, intervals: List[List[int]]) -> int:
    start = [interval[0] for interval in intervals]
    end = [interval[1] for interval in intervals]

    start.sort()
    end.sort()

    minNumofServers = 0
    count = 0
    i = 0
    j = 0

    while i < len(intervals):
        if start[i] < end[j]:
            count += 1
            i += 1
        else:
            count -= 1
            j += 1
        minNumofServers = max(minNumofServers, count)

    return minNumofServers
 

File: ExtraCredit/HW7/HW7.java
import java.util.Arrays;

public class HW7 {
    public int minMeetingRooms(int[][] intervals) {
        int[] start = new int[intervals.length];
        int[] end = new int[intervals.length];

        for (int i = 0; i < intervals.length; i++) {
            start[i] = intervals[i][0];
            end[i] = intervals[i][1];
        }
        Arrays.sort(start);
        Arrays.sort(end);

        int minNumOfServers = 0;
        int count = 0;
        int i = 0; // pointer for start array
        int j = 0; // pointer for end array

        while (i < intervals.length) {
            if (start[i] < end[j]) {
                count++;
                i++;
            } else {
                count--;
                j++;
            }
            minNumOfServers = Math.max(minNumOfServers, count);
        }
        return minNumOfServers;
    }
}
